File: android/app/src/main/java/tech/loveace/appv3/data/model/LaborClubMembership.kt
package tech.loveace.appv3.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonPrimitive

@Serializable
data class LaborClubDirectoryItem(
    @SerialName("ID") val id: String = "",
    @SerialName("Name") val name: String = "",
    @SerialName("TypeID") val typeId: String = "",
    @SerialName("ProjectID") val projectId: String = "",
    @SerialName("PeopleNum") val peopleNum: Int = 0,
    @SerialName("MemberNum") val memberNum: Int = 0,
    @SerialName("PorjectName") val projectName: String = "",
    @SerialName("TypeName") val typeName: String = "",
    @SerialName("Ico") val iconUrl: String? = null,
    @SerialName("Desc") val description: String? = null,
    @SerialName("IsEnable") val isEnabled: Boolean = false,
    @SerialName("IsJoin") val isJoined: Boolean = false,
) {
    val canApply: Boolean get() = isEnabled && !isJoined
}

data class LaborClubApplication(
    val id: String = "",
    val clubId: String = "",
    val clubName: String = "",
    val reason: String = "",
    val addTime: String = "",
    val replyComment: String = "",
    val isAgree: Boolean? = null,
    val statusText: String = "",
) {
    val reviewStatus: LaborClubApplicationReviewStatus
        get() {
            val normalizedStatus = statusText.trim().lowercase()
            if (REJECTED_STATUS_WORDS.any(normalizedStatus::contains)) {
                return LaborClubApplicationReviewStatus.REJECTED
            }
            return when (isAgree) {
                true -> LaborClubApplicationReviewStatus.APPROVED
                false -> LaborClubApplicationReviewStatus.REJECTED
                null -> LaborClubApplicationReviewStatus.PENDING
            }
        }

    private companion object {
        val REJECTED_STATUS_WORDS = listOf("拒绝", "驳回", "未通过", "失效", "过期", "invalid", "expired")
    }
}

enum class LaborClubApplicationReviewStatus {
    PENDING,
    APPROVED,
    REJECTED,
}

enum class LaborClubMembershipStatus {
    JOINED,
    PENDING,
    APPROVED_SYNCING,
    NOT_JOINED,
    REJECTED,
    SUBMITTING,
}

data class LaborClubMembershipState(
    val status: LaborClubMembershipStatus = LaborClubMembershipStatus.NOT_JOINED,
    val latestApplication: LaborClubApplication? = null,
)

data class LaborClubSubmissionResolution(
    val membership: LaborClubMembershipState,
    val isStatusSyncing: Boolean,
)

fun resolveLaborClubMembership(
    joinedClubs: List<LaborClubInfo>,
    latestApplication: LaborClubApplication?,
    isSubmitting: Boolean = false,
): LaborClubMembershipState {
    if (joinedClubs.isNotEmpty()) {
        return LaborClubMembershipState(LaborClubMembershipStatus.JOINED, latestApplication)
    }
    if (isSubmitting) {
        return LaborClubMembershipState(LaborClubMembershipStatus.SUBMITTING, latestApplication)
    }
    val status = when (latestApplication?.reviewStatus) {
        LaborClubApplicationReviewStatus.PENDING -> LaborClubMembershipStatus.PENDING
        LaborClubApplicationReviewStatus.APPROVED -> LaborClubMembershipStatus.APPROVED_SYNCING
        LaborClubApplicationReviewStatus.REJECTED -> LaborClubMembershipStatus.REJECTED
        null -> LaborClubMembershipStatus.NOT_JOINED
    }
    return LaborClubMembershipState(status, latestApplication)
}

fun resolveLaborClubSubmission(
    joinedClubs: List<LaborClubInfo>,
    latestApplication: LaborClubApplication?,
    expectedClubId: String,
    previousApplication: LaborClubApplication? = null,
): LaborClubSubmissionResolution {
    val normalizedClubId = expectedClubId.trim()
    val confirmedApplication = latestApplication
        ?.takeIf { it.clubId.trim().equals(normalizedClubId, ignoreCase = true) }
        ?.takeUnless { previousApplication != null && sameApplicationRecord(it, previousApplication) }
    if (joinedClubs.isNotEmpty()) {
        return LaborClubSubmissionResolution(
            membership = resolveLaborClubMembership(joinedClubs, confirmedApplication),
            isStatusSyncing = false,
        )
    }
    if (confirmedApplication == null) {
        return LaborClubSubmissionResolution(
            membership = LaborClubMembershipState(LaborClubMembershipStatus.PENDING),
            isStatusSyncing = true,
        )
    }
    return LaborClubSubmissionResolution(
        membership = resolveLaborClubMembership(emptyList(), confirmedApplication),
        isStatusSyncing = false,
    )
}

private fun sameApplicationRecord(
    current: LaborClubApplication,
    previous: LaborClubApplication,
): Boolean {
    val currentId = current.id.trim()
    val previousId = previous.id.trim()
    if (currentId.isNotEmpty() && previousId.isNotEmpty()) {
        return currentId.equals(previousId, ignoreCase = true)
    }
    return current.clubId.trim().equals(previous.clubId.trim(), ignoreCase = true) &&
        current.addTime.trim() == previous.addTime.trim() &&
        current.reason.trim() == previous.reason.trim()
}

fun latestLaborClubApplication(applications: List<LaborClubApplication>): LaborClubApplication? =
    applications.maxByOrNull { applicationTimeKey(it.addTime) }

fun decodeLaborClubApplication(element: JsonElement): LaborClubApplication {
    val obj = element as? JsonObject ?: return LaborClubApplication()
    return LaborClubApplication(
        id = obj.stringValue("ID", "Id", "id"),
        clubId = obj.stringValue("ClubID", "ClubId", "clubID", "clubId"),
        clubName = obj.stringValue("ClubName", "Name"),
        reason = obj.stringValue("Reason"),
        addTime = obj.stringValue("AddTime"),
        replyComment = obj.stringValue("ReplyComment", "ExamComment", "Reply"),
        isAgree = obj.flexibleBoolean("IsAgree"),
        statusText = obj.stringValue("StateName", "StatusName", "ApplyStateName"),
    )
}

private fun applicationTimeKey(value: String): String {
    val digits = value.filter(Char::isDigit)
    return digits.padEnd(20, '0').take(20)
}

private fun JsonObject.stringValue(vararg keys: String): String = keys.firstNotNullOfOrNull { key ->
    this[key]?.jsonPrimitive?.contentOrNull
}?.trim().orEmpty()

private fun JsonObject.flexibleBoolean(key: String): Boolean? {
    val element = this[key] ?: return null
    if (element is JsonNull) return null
    val primitive = element.jsonPrimitive
    primitive.booleanOrNull?.let { return it }
    primitive.intOrNull?.let { return it != 0 }
    return when (primitive.contentOrNull?.trim()?.lowercase()) {
        "true", "1", "yes", "approved", "agree", "通过", "同意" -> true
        "false", "0", "no", "rejected", "refused", "拒绝", "驳回", "未通过", "失效" -> false
        else -> null
    }
}

File: ios/loveaceios/Models/LaborClubMembership.swift
import Foundation

struct LaborClubDirectoryItem: Codable, Identifiable {
    let id: String
    let name: String
    let typeId: String
    let projectId: String
    let peopleNum: Int
    let memberNum: Int
    let projectName: String
    let typeName: String
    let iconUrl: String?
    let description: String?
    let isEnabled: Bool
    let isJoined: Bool

    enum CodingKeys: String, CodingKey {
        case id = "ID"
        case name = "Name"
        case typeId = "TypeID"
        case projectId = "ProjectID"
        case peopleNum = "PeopleNum"
        case memberNum = "MemberNum"
        case projectName = "PorjectName"
        case typeName = "TypeName"
        case iconUrl = "Ico"
        case description = "Desc"
        case isEnabled = "IsEnable"
        case isJoined = "IsJoin"
    }

    var canApply: Bool { isEnabled && !isJoined }
}

struct LaborClubApplication: Codable, Identifiable {
    let id: String
    let clubId: String
    let clubName: String
    let reason: String
    let addTime: String
    let replyComment: String
    let isAgree: Bool?
    let statusText: String

    enum CodingKeys: String, CodingKey {
        case id = "ID"
        case clubId = "ClubID"
        case clubName = "ClubName"
        case reason = "Reason"
        case addTime = "AddTime"
        case replyComment = "ReplyComment"
        case isAgree = "IsAgree"
        case statusText = "StateName"
    }

    init(
        id: String = "",
        clubId: String = "",
        clubName: String = "",
        reason: String = "",
        addTime: String = "",
        replyComment: String = "",
        isAgree: Bool? = nil,
        statusText: String = ""
    ) {
        self.id = id
        self.clubId = clubId
        self.clubName = clubName
        self.reason = reason
        self.addTime = addTime
        self.replyComment = replyComment
        self.isAgree = isAgree
        self.statusText = statusText
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decodeIfPresent(String.self, forKey: .id) ?? ""
        clubId = try container.decodeIfPresent(String.self, forKey: .clubId) ?? ""
        clubName = try container.decodeIfPresent(String.self, forKey: .clubName) ?? ""
        reason = try container.decodeIfPresent(String.self, forKey: .reason) ?? ""
        addTime = try container.decodeIfPresent(String.self, forKey: .addTime) ?? ""
        replyComment = try container.decodeIfPresent(String.self, forKey: .replyComment) ?? ""
        statusText = try container.decodeIfPresent(String.self, forKey: .statusText) ?? ""
        isAgree = container.decodeFlexibleBoolIfPresent(forKey: .isAgree)
    }

    var reviewStatus: LaborClubApplicationReviewStatus {
        let normalized = statusText.lowercased()
        let rejectedWords = ["拒绝", "驳回", "未通过", "失效", "过期", "invalid", "expired"]
        if rejectedWords.contains(where: normalized.contains) { return .rejected }
        switch isAgree {
        case true: return .approved
        case false: return .rejected
        case nil: return .pending
        }
    }
}

enum LaborClubApplicationReviewStatus {
    case pending
    case approved
    case rejected
}

enum LaborClubMembershipStatus {
    case joined
    case pending
    case approvedSyncing
    case notJoined
    case rejected
    case submitting
}

struct LaborClubMembershipState {
    let status: LaborClubMembershipStatus
    let latestApplication: LaborClubApplication?

    static let notJoined = LaborClubMembershipState(status: .notJoined, latestApplication: nil)
}

struct LaborClubSubmissionResolution {
    let membership: LaborClubMembershipState
    let isStatusSyncing: Bool
}

func resolveLaborClubMembership(
    joinedClubs: [LaborClubInfo],
    latestApplication: LaborClubApplication?,
    isSubmitting: Bool = false
) -> LaborClubMembershipState {
    if !joinedClubs.isEmpty {
        return LaborClubMembershipState(status: .joined, latestApplication: latestApplication)
    }
    if isSubmitting {
        return LaborClubMembershipState(status: .submitting, latestApplication: latestApplication)
    }
    let status: LaborClubMembershipStatus
    switch latestApplication?.reviewStatus {
    case .pending: status = .pending
    case .approved: status = .approvedSyncing
    case .rejected: status = .rejected
    case nil: status = .notJoined
    }
    return LaborClubMembershipState(status: status, latestApplication: latestApplication)
}

func resolveLaborClubSubmission(
    joinedClubs: [LaborClubInfo],
    latestApplication: LaborClubApplication?,
    expectedClubId: String,
    previousApplication: LaborClubApplication? = nil
) -> LaborClubSubmissionResolution {
    let normalizedClubId = expectedClubId.trimmingCharacters(in: .whitespacesAndNewlines)
    let confirmedApplication: LaborClubApplication? = latestApplication.flatMap { application -> LaborClubApplication? in
        guard application.clubId.trimmingCharacters(in: .whitespacesAndNewlines)
            .caseInsensitiveCompare(normalizedClubId) == .orderedSame else { return nil }
        if let previousApplication, sameApplicationRecord(application, previousApplication) {
            return nil
        }
        return application
    }
    if !joinedClubs.isEmpty {
        return LaborClubSubmissionResolution(
            membership: resolveLaborClubMembership(
                joinedClubs: joinedClubs,
                latestApplication: confirmedApplication
            ),
            isStatusSyncing: false
        )
    }
    guard let confirmedApplication else {
        return LaborClubSubmissionResolution(
            membership: LaborClubMembershipState(status: .pending, latestApplication: nil),
            isStatusSyncing: true
        )
    }
    return LaborClubSubmissionResolution(
        membership: resolveLaborClubMembership(
            joinedClubs: [],
            latestApplication: confirmedApplication
        ),
        isStatusSyncing: false
    )
}

private func sameApplicationRecord(
    _ current: LaborClubApplication,
    _ previous: LaborClubApplication
) -> Bool {
    let currentId = current.id.trimmingCharacters(in: .whitespacesAndNewlines)
    let previousId = previous.id.trimmingCharacters(in: .whitespacesAndNewlines)
    if !currentId.isEmpty, !previousId.isEmpty {
        return currentId.caseInsensitiveCompare(previousId) == .orderedSame
    }
    return current.clubId.trimmingCharacters(in: .whitespacesAndNewlines)
        .caseInsensitiveCompare(previous.clubId.trimmingCharacters(in: .whitespacesAndNewlines)) == .orderedSame &&
        current.addTime.trimmingCharacters(in: .whitespacesAndNewlines) ==
            previous.addTime.trimmingCharacters(in: .whitespacesAndNewlines) &&
        current.reason.trimmingCharacters(in: .whitespacesAndNewlines) ==
            previous.reason.trimmingCharacters(in: .whitespacesAndNewlines)
}

func latestLaborClubApplication(_ applications: [LaborClubApplication]) -> LaborClubApplication? {
    applications.max { applicationTimeKey($0.addTime) < applicationTimeKey($1.addTime) }
}

private func applicationTimeKey(_ value: String) -> String {
    let digits = value.filter(\.isNumber)
    return String((digits + String(repeating: "0", count: 20)).prefix(20))
}

private extension KeyedDecodingContainer {
    func decodeFlexibleBoolIfPresent(forKey key: Key) -> Bool? {
        if (try? decodeNil(forKey: key)) == true { return nil }
        if let value = try? decode(Bool.self, forKey: key) { return value }
        if let value = try? decode(Int.self, forKey: key) { return value != 0 }
        guard let value = try? decode(String.self, forKey: key) else { return nil }
        switch value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
        case "true", "1", "yes", "approved", "agree", "通过", "同意": return true
        case "false", "0", "no", "rejected", "refused", "拒绝", "驳回", "未通过", "失效": return false
        default: return nil
        }
    }
}
